### PR TITLE
Modify error fomat

### LINKED ISSUE #2: Better error indication
```erg
f x =
    ...
    y = foo.try_some(x)?
    ...

g x =
    y = f(x)?
    ...

i = g(1)?
```

Suppose you get the following error in the above code (including omissions).

```console
Traceback (most recent call first):
    ...
    Foo.try_some, line ..., file "..."
    .... | y = foo.try_some(x)?
    module::f, line ..., file "..."
    .... | y = f(x)?
    module::g, line ..., file "..."
    .... | i = g(1)?
Error: ...
```

How about displaying the contents and type of the variable to make the error message clearer?
Like this:

```console
Traceback (most recent call first):
    ...
    Foo.try_some, line ..., file "..."
    .... | y = foo.try_some(x)?
               │            └ 1: Int
               └ <Foo object>: Foo
    module::f, line ..., file "..."
    ... | y = f(x)?
                └ 1: Int
    module::g, line ..., file "..."
    ... | i = g(1)?
Error: ...
```

### PR DESCRIPTION
**Pull request by GreasySlug**
Fix: #2.

## Intro

The error handling code you wrote was very clear and abstract.
But what I am rewriting now is not.
This code is pretty dirty and you can just discard it.

## Changes proposed in this PR

- [x] Add styled about color and attribute
- [x] Coloring using prescribed colors
- [x] Modify error formatting using prescribed color and chars
- [ ] Fix #2 
- [ ] Long error description formatting

I want to reformat like following.

- Current format
```python
+----------------+
| header         |
+----------------+
| linen + codes  |
|         points |
+----------------+
| footer         |
+----------------+
Error[#0555]: File .\examples\array.er, line 6, in <module>
6│ mut_content_arr.map!(x -> x + 1)
   ^^^^^^^^^^^^^^^
NameError: ::mut_content_arr(: Failure) is not defined
Error[#0555]: File .\examples\array.er, line 10, in <module>
10│ telescoping_arr.push!(6)
    ^^^^^^^^^^^^^^^
NameError: ::telescoping_arr(: Failure) is not defined
```
- [WIP] Modifying format
```python
+-------------+
| header      |
+-------------+
| codes       |
| lbot + desc |
+ ------------+
| footer      |
+-------------+
Error[#0572]: File .\examples\array.er, line 6, in <module>

6 | mut_content_arr.map!(x -> x + 1)
  : ---------------
  :                `- NameError: ::mut_content_arr(: Failure) is not defined

Error[#0572]: File .\examples\array.er, line 10, in <module>

10 | telescoping_arr.push!(6)
   : ---------------
   :                `- NameError: ::telescoping_arr(: Failure) is not define

Error[#0357]: File <stdin>, line 1, in <module>

1 | print
  : -----
  :      `- NameError: print is not defined

hint: exists a similar name variable: print!
```

## Edit

Enable colors and chars in a specified theme experimentally.

```shell
cargo run --features "unicode pretty"
```

Default is ANSI color and ASCII.

The above will use unicode chars and custom colors.

- `unicode`: use `╰`, `─`, `·`  and `│`
- `pretty`: use emoji and custom colors
    This is a experimental addition to prove that changes can be made
    Custom colors are put on in order to set up coloring for CVD(Color Vision Deficiency) etc. in the future, and will be separated from `pretty`

- original error format
origin_err

- default format
default_err

- unicode and pretty format
emoji_err

@mtshiba

### CODE REVIEW

**Comment by GreasySlug** [comment]
When error position is on the right side and the error description sentence is long, the whole sentence becomes longer to the right.
```python
4 │ mut_arr = [1, 2, 3, 4, 5].into [Int!; !5]
                             -----
                                 ╰─ AttributeError: Array({1, 4, 3, 5, 2, }, 5) object has no attribute into
```
Specify the width to align these to the left.
```python
4 │ mut_arr = [1, 2, 3, 4, 5].into [Int!; !5]
  ╭──────────────────────────-----
  ╰─ AttributeError: Array({1, 4, 3, 5, 2, }, 5) object has no attribute into
```

**Comment by C-BJ** [comment]
Consider direct fix #2? @GreasySlug

**Comment by GreasySlug** [comment]
No, color and formats only.
I' m trying to make it easier for #2 to color and modify as needed with this format.
I'm in the process of looking into several error handling libraries to do this.

**Comment by C-BJ** [comment]
> ~No, color and formats only~ Maybe we should work on this.. I'm trying to make it easier for #2 to color and modify as needed with this format. I'm in the process of looking into several error handling libraries to do this.

#2 will be closed if this PR merges
image

**Comment by GreasySlug** [comment]
> When error position is on the right side and the error description sentence is long, the whole sentence becomes longer to the right.
> 
> ```python
> 4 │ mut_arr = [1, 2, 3, 4, 5].into [Int!; !5]
>                              -----
>                                  ╰─ AttributeError: Array({1, 4, 3, 5, 2, }, 5) object has no attribute into
> ```
> 
> Specify the width to align these to the left.
> 
> ```python
> 4 │ mut_arr = [1, 2, 3, 4, 5].into [Int!; !5]
>   ╭──────────────────────────-----
>   ╰─ AttributeError: Array({1, 4, 3, 5, 2, }, 5) object has no attribute into
> ```

Same as #2 on this, we need more than one opinion.
Adding a type will inevitably lengthen it.
User type may cause errors due to very long types.
Then the format can't handle it.

**Comment by mtshiba** [comment]
That is OK for now.

Ultimately, the error message is split into two parts: the main message and the submessage.
The sub-message is simple and appears directly below the code block.
The main message is longer and more detailed and appears at the end.

**Comment by GreasySlug** [comment]
> That is OK for now.
> 
> Ultimately, the error message is split into two parts: the main message and the submessage. The submessage is simple and appears directly below the code block. The main message is longer and more detailed and appears at the end.
> 
> The format will be like:
> 
> ```
> n | invalid_code
>     ____________
>                ╰─ sub message
> ErrorKind: main message
> ```

That's good to see.
How do you plan to separate error messages into sub and main messages?

File: compiler/erg_common/style.rs [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+        }
+    }
+}
+
+///
+/// `StringSpan` is for coloring and attribute text.
+/// String, Color(&str) and Attribute(&str)
+///
+#[derive(Debug)]
+pub struct StringSpan {
```
Comment (mtshiba) [inline compiler/erg_common/style.rs:292]: I think the mean of `Dpan` is unclear.
Wouldn't `DecoratedString` be better?

Comment (mtshiba) [reply compiler/erg_common/style.rs:292]: Or `StyledString`.

**Comment by mtshiba** [comment]
> That's good to see. How do you plan to separate error messages into sub and main messages?

This changes also involve displaying errors across multiple code blocks. Let's merge this PR and then open it as a separate PR.

File: compiler/erg_common/style.rs (hunk 1 above)
Comment (GreasySlug) [reply compiler/erg_common/style.rs:292]: This `Span` comes from [html's span](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span)
The word Span has no meaning and the word `Span` itself means modified text

Comment (mtshiba) [reply compiler/erg_common/style.rs:292]: I think `StringSpan` can be misinterpreted as a string having information about its position and span in the code.

Comment (GreasySlug) [reply compiler/erg_common/style.rs:292]: I certainly didn't take that into account.
I'll change `StringSpan' to `StyledString`

**Pull request merged**